src/webapi/rendering_context.rs

```diff
@@ -27,7 +190,809 @@ impl RenderingContext for CanvasRenderingContext2d {
+            }    
+        }
+        else {
+            js! { @(no_return)
+                @{&self.0}.clip();
+            }
+        }
+    }
+
+    /// Causes the point of the pen to move back to the start of the current sub-path. It tries 
+    /// to add a straight line (but does not actually draw it) from the current point to the start. 
+    /// If the shape has already been closed or has only one point, this function does nothing.
+    /// 
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/closePath)
+    // https://html.spec.whatwg.org/#2dcontext:dom-context-2d-closepath
+    pub fn close_path(&self) {
+        js! { @(no_return)
+            @{&self.0}.closePath();
+        }
+    }
+
+    /// Creates a gradient along the line given by the coordinates represented by the parameters.
+    /// 
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/createLinearGradient)
+    // https://html.spec.whatwg.org/#2dcontext:dom-context-2d-createlineargradient
+    pub fn create_linear_gradient(&self, x0: f64, y0: f64, x1: f64, y1: f64) -> CanvasGradient {
+        js! (
+            return @{&self.0}.createLinearGradient(@{x0}, @{y0}, @{x1}, @{y1});
+        ).try_into().unwrap()
+    }
+
+    /// Creates a new, blank ImageData object with the specified dimensions. 
+    /// All of the pixels in the new object are transparent black.
+    /// 
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/createImageData)
+    // https://html.spec.whatwg.org/#2dcontext:dom-context-2d-createimagedata
+    pub fn create_image_data_wh(&self, width: f64, height: f64) -> Result<ImageData, IndexSizeError> {
+        js_try! (
+            return @{&self.0}.createImageData(@{width}, @{height});
+        ).unwrap()
+    }
+
+    /// Creates a new, blank ImageData object with the specified dimensions. 
+    /// All of the pixels in the new object are transparent black.
+    /// 
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/createImageData)
+    // https://html.spec.whatwg.org/#2dcontext:dom-context-2d-createimagedata
+    pub fn create_image_data_id(&self, image_data: ImageData) -> ImageData {
+        js! (
+            return @{&self.0}.createImageData(@{image_data});
+        ).try_into().unwrap()
+    }
+
+    /// Creates a pattern using the specified image (a CanvasImageSource). It repeats the source in 
+    /// the directions specified by the repetition argument. This method returns a CanvasPattern.
+    /// 
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/createPattern)
+    // https://html.spec.whatwg.org/#2dcontext:dom-context-2d-createpattern
+    pub fn create_pattern_image(&self, image: ImageElement, repetition: Option<Repetition>) -> Result<CanvasPattern, SyntaxError> {
```

Comment:
Also, since the `Repetition` has a well defined default value perhaps it would be a good idea to implement `Default` for it and make this method take `Repetition` instead of `Option<Repetition>`? (Then you would call it `create_pattern_image(image, Repetition::default())`.)

Ditto for methods which take a `FillRule` (where fill rule is always "nonzero" by default from what I can see.)